File: tools/context_artifact_builder_v1/extract.py
from __future__ import annotations

import re
from pathlib import Path
from typing import Any, Mapping, Sequence

from ticket_issuance_planner_v2.context import (
    validate_context,
    validate_control_schema,
    validate_registries,
)
from ticket_issuance_planner_v2.control import (
    validate_handoffs,
    validate_root_metadata_and_package_state,
    validate_workflows,
)
from ticket_issuance_planner_v2.core import (
    Checks,
    DraftPair,
    GitView,
    PlannerFailure,
    canonical_json_bytes,
    exact_sha256,
    one_table,
)
from ticket_issuance_planner_v2.drafts import load_draft_pair

from .core import (
    ARTIFACT_FORMAT,
    ARTIFACT_ROOT,
    RECORD_KIND,
    CandidateFailure,
    BundleBlock,
    map_planner_exception,
    normalize_utf8_lf,
    planner_failure_from_checks,
    require_json_value,
)


def validate_builder_contract(view: GitView) -> None:
    try:
        registry, _ = view.load_toml("swarm/context-artifact-builder-v1.toml")
        schema, _ = view.load_toml("swarm/context-artifact-candidate-schema-v1.toml")
        digest, _ = view.load_toml("swarm/context-artifact-candidate-digest-v1.toml")
    except PlannerFailure as exc:
        raise map_planner_exception(exc) from exc
    authority = registry.get("authority")
    coherent = (
        registry.get("schema_version") == 1
        and registry.get("component") == "context_artifact_builder_v1"
        and registry.get("candidate_schema")
        == "swarm/context-artifact-candidate-schema-v1.toml"
        and registry.get("digest_profile")
        == "swarm/context-artifact-candidate-digest-v1.toml"
        and registry.get("artifact_root") == ARTIFACT_ROOT
        and registry.get("artifact_format") == ARTIFACT_FORMAT
        and schema.get("schema_version") == 1
        and schema.get("record_kind") == RECORD_KIND
        and schema.get("artifact_format") == ARTIFACT_FORMAT
        and digest.get("schema_version") == 1
        and digest.get("profile") == "context_artifact_candidate_digest_v1"
        and digest.get("self_referential_digest_allowed") is False
        and isinstance(authority, dict)
        and authority
        and all(value is False for value in authority.values())
    )
    if not coherent:
        raise CandidateFailure(
            "CONTEXT_ARTIFACT_BUILDER_CONTRACT_MISMATCH",
            "context artifact builder registry/schema/digest contract mismatch",
        )


def preflight(
    root: str,
    package: str,
    base_commit: str,
    accepted_handoff_paths: Sequence[str],
) -> tuple[
    GitView,
    DraftPair,
    Mapping[str, Any],
    Mapping[str, Any],
    list[dict[str, Any]],
    list[dict[str, str]],
]:
    try:
        view = GitView(Path(root), base_commit)
    except PlannerFailure as exc:
        raise map_planner_exception(exc) from exc
    validate_builder_contract(view)

    checks = Checks()
    launch, package_row, _function_row, _stage_row = validate_registries(
        view, package, checks
    )
    pair = load_draft_pair(view, package, package_row, checks)
    if pair is not None:
        validate_context(view, pair, package, checks)
    validate_control_schema(view, launch, checks)
    validate_root_metadata_and_package_state(view, package, checks)
    validate_workflows(view, checks)
    handoffs = (
        validate_handoffs(view, pair, accepted_handoff_paths, checks)
        if pair is not None
        else []
    )
    if pair is None or checks.reasons:
        raise planner_failure_from_checks(checks.items)
    assert package_row is not None
    classification = (
        "AUTHORIZED"
        if package in launch.get("authorized_packages", [])
        else "CONDITIONAL"
        if package in launch.get("conditional_packages", [])
        else "UNKNOWN"
    )
    if pair.ticket.get("launch_class") != classification or classification == "UNKNOWN":
        raise CandidateFailure(
            "PACKAGE_STAGE_MISMATCH",
            "draft and launch classification differ",
            checks.items,
        )
    return view, pair, launch, package_row, handoffs, checks.items


def source_records_and_blocks(
    view: GitView, pair: DraftPair
) -> tuple[list[dict[str, Any]], list[BundleBlock]]:
    canonicalization = pair.context.get("canonicalization")
    if not isinstance(canonicalization, dict):
        raise CandidateFailure("DRAFT_PAIR_MISMATCH", "missing context canonicalization")
    if canonicalization.get("path_header_format") != "--- repository-path: <path> ---":
        raise CandidateFailure(
            "DRAFT_PAIR_MISMATCH", "source header format is not canonical"
        )
    records: list[dict[str, Any]] = []
    blocks: list[BundleBlock] = []
    for order, path in enumerate(pair.sources):
        try:
            raw, entry = view.read_bytes(path)
        except PlannerFailure as exc:
            raise map_planner_exception(exc) from exc
        normalized = normalize_utf8_lf(raw, path)
        record = {
            "order": order,
            "repository_path": path,
            "git_blob_id": view.blob_identity(entry),
            "exact_sha256": exact_sha256(raw),
            "exact_bytes": len(raw),
            "materialization": "UTF8_LF",
            "materialized_sha256": exact_sha256(normalized),
            "materialized_bytes": len(normalized),
        }
        records.append(record)
        blocks.append(
            BundleBlock(
                "source",
                f"--- repository-path: {path} ---",
                record,
                normalized,
            )
        )
    return records, blocks


def _selector_value(
    view: GitView, selector: str, package: str
) -> tuple[str, str, Any, str, str]:
    if "::" not in selector:
        raise CandidateFailure("CONTEXT_SELECTOR_INVALID", f"invalid selector: {selector}")
    path, expression = selector.split("::", 1)
    try:
        document, entry = view.load_toml(path)
        source_raw, _ = view.read_bytes(path)
    except PlannerFailure as exc:
        raise map_planner_exception(exc) from exc

    value: Any = None
    match = re.fullmatch(r"package\[name=([a-z][a-z0-9-]*)\]", expression)
    if match and path == "swarm/crates.toml" and match.group(1) == package:
        value = one_table(document.get("package"), "name", package)
    match = re.fullmatch(r"foundation\[package=([a-z][a-z0-9-]*)\]", expression)
    if match and path == "swarm/function-packets.toml" and match.group(1) == package:
        value = one_table(document.get("foundation"), "package", package)
    match = re.fullmatch(r"stage\[id=(W(?:10|[0-9]))\]", expression)
    if match and path == "swarm/stages.toml" and match.group(1) == "W0":
        value = one_table(document.get("stage"), "id", "W0")
    match = re.fullmatch(
        r"(authorized_packages|conditional_packages)\[([a-z][a-z0-9-]*)\]",
        expression,
    )
    if match and path == "swarm/launch-state.toml" and match.group(2) == package:
        values = document.get(match.group(1))
        if isinstance(values, list) and values.count(package) == 1:
            value = {"membership": match.group(1), "package": package}
        else:
            value = None
    match = re.fullmatch(r"conditional_activation\.([a-z][a-z0-9-]*)", expression)
    if match and path == "swarm/launch-state.toml" and match.group(1) == package:
        table = document.get("conditional_activation")
        value = table.get(package) if isinstance(table, dict) else None

    if value is None:
        raise CandidateFailure(
            "CONTEXT_SELECTOR_NOT_UNIQUE",
            f"selector did not resolve exactly once: {selector}",
        )
    require_json_value(value, selector)
    return (
        path,
        expression,
        value,
        view.blob_identity(entry),
        exact_sha256(source_raw),
    )


def fragment_records_and_blocks(
    view: GitView, pair: DraftPair, package: str
) -> tuple[list[dict[str, Any]], list[BundleBlock]]:
    canonicalization = pair.context.get("canonicalization")
    if not isinstance(canonicalization, dict):
        raise CandidateFailure("DRAFT_PAIR_MISMATCH", "missing context canonicalization")
    if (
        canonicalization.get("registry_header_format")
        != "--- registry-selector: <path>::<selector> ---"
    ):
        raise CandidateFailure(
            "DRAFT_PAIR_MISMATCH", "registry header format is not canonical"
        )
    records: list[dict[str, Any]] = []
    blocks: list[BundleBlock] = []
    for order, selector in enumerate(pair.selectors):
        path, expression, value, blob_id, source_sha = _selector_value(
            view, selector, package
        )
        fragment = canonical_json_bytes(
            {"registry_path": path, "selector": expression, "value": value}
        )
        record = {
            "order": order,
            "registry_path": path,
            "selector": expression,
            "source_git_blob_id": blob_id,
            "source_exact_sha256": source_sha,
            "selector_match_count": 1,
            "fragment_sha256": exact_sha256(fragment),
            "fragment_bytes": len(fragment),
        }
        records.append(record)
        blocks.append(
            BundleBlock(
                "registry_fragment",
                f"--- registry-selector: {path}::{expression} ---",
                record,
                fragment,
            )
        )
    return records, blocks


def handoff_records_and_blocks(
    view: GitView, handoffs: Sequence[Mapping[str, Any]]
) -> tuple[list[dict[str, Any]], list[BundleBlock]]:
    records: list[dict[str, Any]] = []
    blocks: list[BundleBlock] = []
    for order, summary in enumerate(handoffs):
        path = str(summary["path"])
        try:
            raw, _entry = view.read_bytes(path)
            text = raw.decode("utf-8", "strict")
        except (PlannerFailure, UnicodeDecodeError) as exc:
            if isinstance(exc, PlannerFailure):
                raise map_planner_exception(exc) from exc
            raise CandidateFailure(
                "HANDOFF_RECORD_INVALID", f"handoff is not UTF-8: {path}"
            ) from exc
        if "\r" in text or not raw.endswith(b"\n"):
            raise CandidateFailure(
                "HANDOFF_RECORD_INVALID",
                f"handoff is not exact UTF-8/LF with terminal LF: {path}",
            )
        record = dict(summary)
        record.update(
            {
                "order": order,
                "materialization": "EXACT_UTF8_LF",
                "materialized_sha256": exact_sha256(raw),
                "materialized_bytes": len(raw),
            }
        )
        records.append(record)
        blocks.append(
            BundleBlock(
                "accepted_handoff",
                f"--- accepted-handoff: {summary['package']} ---",
                record,
                raw,
            )
        )
    return records, blocks

File: xtask/src/context_artifact_builder/extract.rs
//! Immutable-tree source, registry-fragment and handoff extraction.

use serde_json::{Map as JsonMap, Number as JsonNumber, Value as JsonValue, json};
use toml::Value as TomlValue;

use crate::context_artifact::{
    BundleBlock, expected_header, normalize_utf8_lf, require_json_value,
};
use crate::ticket_planner::{canonical_json_bytes, exact_sha256_hex};

use super::model::{ContextArtifactBuildError, Preflight};

/// Fully extracted semantic records plus their ordered bundle blocks.
#[derive(Clone, Debug)]
pub(super) struct Extracted {
    pub(super) sources: Vec<JsonValue>,
    pub(super) fragments: Vec<JsonValue>,
    pub(super) handoffs: Vec<JsonValue>,
    pub(super) blocks: Vec<BundleBlock>,
}

/// Extracts every candidate input from the immutable Git view selected during
/// preflight. The working tree is never consulted.
pub(super) fn extract(
    preflight: &Preflight,
    package: &str,
) -> Result<Extracted, ContextArtifactBuildError> {
    let (sources, source_blocks) = source_records(preflight)?;
    let (fragments, fragment_blocks) = fragment_records(preflight, package)?;
    let (handoffs, handoff_blocks) = handoff_records(preflight)?;
    let mut blocks = Vec::with_capacity(
        source_blocks.len() + fragment_blocks.len() + handoff_blocks.len(),
    );
    blocks.extend(source_blocks);
    blocks.extend(fragment_blocks);
    blocks.extend(handoff_blocks);
    Ok(Extracted {
        sources,
        fragments,
        handoffs,
        blocks,
    })
}

fn source_records(
    preflight: &Preflight,
) -> Result<(Vec<JsonValue>, Vec<BundleBlock>), ContextArtifactBuildError> {
    let mut records = Vec::with_capacity(preflight.pair.sources.len());
    let mut blocks = Vec::with_capacity(preflight.pair.sources.len());
    for (order, path) in preflight.pair.sources.iter().enumerate() {
        let (raw, entry) = preflight
            .tree
            .read_bytes(path)
            .map_err(map_git)?;
        let normalized = normalize_utf8_lf(&raw).map_err(map_primitive)?;
        let record = json!({
            "order": order,
            "repository_path": path,
            "git_blob_id": preflight.tree.blob_identity(&entry),
            "exact_sha256": exact_sha256_hex(&raw),
            "exact_bytes": raw.len(),
            "materialization": "UTF8_LF",
            "materialized_sha256": exact_sha256_hex(&normalized),
            "materialized_bytes": normalized.len(),
        });
        let header = expected_header("source", &record).map_err(map_primitive)?;
        blocks.push(BundleBlock {
            kind: "source".to_owned(),
            header,
            metadata: record.clone(),
            content: normalized,
        });
        records.push(record);
    }
    Ok((records, blocks))
}

fn fragment_records(
    preflight: &Preflight,
    package: &str,
) -> Result<(Vec<JsonValue>, Vec<BundleBlock>), ContextArtifactBuildError> {
    let mut records = Vec::with_capacity(preflight.pair.selectors.len());
    let mut blocks = Vec::with_capacity(preflight.pair.selectors.len());
    for (order, selector) in preflight.pair.selectors.iter().enumerate() {
        let (path, expression) = selector.split_once("::").ok_or_else(|| {
            ContextArtifactBuildError::new(
                "CONTEXT_SELECTOR_INVALID",
                format!("invalid selector: {selector}"),
            )
        })?;
        let (document, entry) = preflight.tree.load_toml(path).map_err(map_git)?;
        let (source_raw, _) = preflight.tree.read_bytes(path).map_err(map_git)?;
        let value = resolve_selector(&document, path, expression, package)?;
        require_json_value(&value).map_err(map_primitive)?;
        let fragment = canonical_json_bytes(&json!({
            "registry_path": path,
            "selector": expression,
            "value": value,
        }));
        let record = json!({
            "order": order,
            "registry_path": path,
            "selector": expression,
            "source_git_blob_id": preflight.tree.blob_identity(&entry),
            "source_exact_sha256": exact_sha256_hex(&source_raw),
            "selector_match_count": 1,
            "fragment_sha256": exact_sha256_hex(&fragment),
            "fragment_bytes": fragment.len(),
        });
        let header = expected_header("registry_fragment", &record)
            .map_err(map_primitive)?;
        blocks.push(BundleBlock {
            kind: "registry_fragment".to_owned(),
            header,
            metadata: record.clone(),
            content: fragment,
        });
        records.push(record);
    }
    Ok((records, blocks))
}

fn handoff_records(
    preflight: &Preflight,
) -> Result<(Vec<JsonValue>, Vec<BundleBlock>), ContextArtifactBuildError> {
    let mut records = Vec::with_capacity(preflight.handoffs.len());
    let mut blocks = Vec::with_capacity(preflight.handoffs.len());
    for (order, handoff) in preflight.handoffs.iter().enumerate() {
        let text = std::str::from_utf8(&handoff.bytes).map_err(|_| {
            ContextArtifactBuildError::new(
                "HANDOFF_RECORD_INVALID",
                "handoff is not UTF-8",
            )
        })?;
        if text.contains('\r') || !handoff.bytes.ends_with(b"\n") {
            return Err(ContextArtifactBuildError::new(
                "HANDOFF_RECORD_INVALID",
                "handoff is not exact UTF-8/LF with terminal LF",
            ));
        }
        let mut record = handoff
            .summary
            .as_object()
            .cloned()
            .ok_or_else(|| {
                ContextArtifactBuildError::new(
                    "HANDOFF_RECORD_INVALID",
                    "handoff summary is not an object",
                )
            })?;
        record.insert(
            "order".to_owned(),
            JsonValue::Number(JsonNumber::from(
                u64::try_from(order).unwrap_or(u64::MAX),
            )),
        );
        record.insert(
            "materialization".to_owned(),
            JsonValue::String("EXACT_UTF8_LF".to_owned()),
        );
        record.insert(
            "materialized_sha256".to_owned(),
            JsonValue::String(exact_sha256_hex(&handoff.bytes)),
        );
        record.insert(
            "materialized_bytes".to_owned(),
            JsonValue::Number(JsonNumber::from(
                u64::try_from(handoff.bytes.len()).unwrap_or(u64::MAX),
            )),
        );
        let record = JsonValue::Object(record);
        let header = expected_header("accepted_handoff", &record)
            .map_err(map_primitive)?;
        blocks.push(BundleBlock {
            kind: "accepted_handoff".to_owned(),
            header,
            metadata: record.clone(),
            content: handoff.bytes.clone(),
        });
        records.push(record);
    }
    Ok((records, blocks))
}

fn resolve_selector(
    document: &TomlValue,
    path: &str,
    expression: &str,
    package: &str,
) -> Result<JsonValue, ContextArtifactBuildError> {
    if let Some(name) = bracket_value(expression, "package[name=") {
        if name == package
            && matches!(path, "swarm/crates.toml" | "swarm/modules/w0.toml")
        {
            return unique_toml_row(document, "package", "name", package)
                .and_then(toml_to_json);
        }
    }
    if let Some(name) = bracket_value(expression, "foundation[package=") {
        if name == package && path == "swarm/function-packets.toml" {
            return unique_toml_row(document, "foundation", "package", package)
                .and_then(toml_to_json);
        }
    }
    if let Some(stage) = bracket_value(expression, "stage[id=") {
        if stage == "W0" && path == "swarm/stages.toml" {
            return unique_toml_row(document, "stage", "id", "W0")
                .and_then(toml_to_json);
        }
    }
    for membership in ["authorized_packages", "conditional_packages"] {
        let prefix = format!("{membership}[");
        if expression
            .strip_prefix(&prefix)
            .and_then(|tail| tail.strip_suffix(']'))
            == Some(package)
            && path == "swarm/launch-state.toml"
        {
            let matches = document
                .get(membership)
                .and_then(TomlValue::as_array)
                .map(|values| {
                    values
                        .iter()
                        .filter(|value| value.as_str() == Some(package))
                        .count()
                })
                .unwrap_or_default();
            if matches == 1 {
                return Ok(json!({"membership": membership, "package": package}));
            }
        }
    }
    if expression == format!("conditional_activation.{package}")
        && path == "swarm/launch-state.toml"
    {
        if let Some(value) = document
            .get("conditional_activation")
            .and_then(TomlValue::as_table)
            .and_then(|table| table.get(package))
        {
            return toml_to_json(value.clone());
        }
    }
    Err(ContextArtifactBuildError::new(
        "CONTEXT_SELECTOR_NOT_UNIQUE",
        format!("selector did not resolve exactly once: {path}::{expression}"),
    ))
}

fn bracket_value<'a>(expression: &'a str, prefix: &str) -> Option<&'a str> {
    expression.strip_prefix(prefix)?.strip_suffix("]")
}

fn unique_toml_row(
    document: &TomlValue,
    array: &str,
    key: &str,
    expected: &str,
) -> Result<TomlValue, ContextArtifactBuildError> {
    let rows = document
        .get(array)
        .and_then(TomlValue::as_array)
        .map(|rows| {
            rows.iter()
                .filter(|row| row.get(key).and_then(TomlValue::as_str) == Some(expected))
                .cloned()
                .collect::<Vec<_>>()
        })
        .unwrap_or_default();
    if rows.len() == 1 {
        Ok(rows.into_iter().next().expect("one row"))
    } else {
        Err(ContextArtifactBuildError::new(
            "CONTEXT_SELECTOR_NOT_UNIQUE",
            format!("{array}[{key}={expected}] did not resolve exactly once"),
        ))
    }
}

fn toml_to_json(value: TomlValue) -> Result<JsonValue, ContextArtifactBuildError> {
    match value {
        TomlValue::String(value) => Ok(JsonValue::String(value)),
        TomlValue::Integer(value) => Ok(JsonValue::Number(JsonNumber::from(value))),
        TomlValue::Boolean(value) => Ok(JsonValue::Bool(value)),
        TomlValue::Array(values) => values
            .into_iter()
            .map(toml_to_json)
            .collect::<Result<Vec<_>, _>>()
            .map(JsonValue::Array),
        TomlValue::Table(values) => {
            let mut map = JsonMap::new();
            for (key, value) in values {
                map.insert(key, toml_to_json(value)?);
            }
            Ok(JsonValue::Object(map))
        }
        TomlValue::Float(_) | TomlValue::Datetime(_) => {
            Err(ContextArtifactBuildError::new(
                "REGISTRY_FRAGMENT_NONCANONICAL",
                "selector contains a forbidden float or datetime value",
            ))
        }
    }
}

fn map_git(error: crate::git_tree::GitTreeError) -> ContextArtifactBuildError {
    ContextArtifactBuildError::new(error.reason(), error.message())
}

fn map_primitive(
    error: crate::context_artifact::ContextArtifactError,
) -> ContextArtifactBuildError {
    ContextArtifactBuildError::new(error.reason(), error.message())
}
